File: test/church.lua
local church = {}

local function zero()
    return function(x) return x end
end

local function succ(c)
    return function(f)
        return function(x)
            return f(c(f)(x))
        end
    end
end

local function exp(c, e)
    return e(c)
end

local function tochurch(n)
    if n == 0 then
        return zero
    else
        return succ(tochurch(n-1))
    end
end

local function tolua(c)
    return c(function (x) return x + 1 end)(0)
end

print(tolua(exp(tochurch(2), tochurch(24))))

File: test/church.js

const zero = () => x => x;
const succ = c => f => x => f(c(f)(x));
const exp = (c, e) => e(c);
const tochurch = n => n === 0 ? zero : succ(tochurch(n-1));
const tojs = (c) => c(x => x + 1)(0);

console.log(tojs(exp(tochurch(2), tochurch(24))));
